In [1]:
import numpy as np
import pandas as pd

**1.** Print the Pandas Version

In [2]:
print(pd.__version__)

2.2.3


**2.** Create a dataframe with three columns with Zero values

In [3]:
dtype = [('C1','int32'), ('C2','float32'), ('C3','float32')]
vals = np.zeros(20, dtype=dtype)
index = ['Row'+str(i) for i in range(1,len(vals)+1)]


df = pd.DataFrame(vals)
print(df)

df = pd.DataFrame(vals, index=index)
print(df)

    C1   C2   C3
0    0  0.0  0.0
1    0  0.0  0.0
2    0  0.0  0.0
3    0  0.0  0.0
4    0  0.0  0.0
5    0  0.0  0.0
6    0  0.0  0.0
7    0  0.0  0.0
8    0  0.0  0.0
9    0  0.0  0.0
10   0  0.0  0.0
11   0  0.0  0.0
12   0  0.0  0.0
13   0  0.0  0.0
14   0  0.0  0.0
15   0  0.0  0.0
16   0  0.0  0.0
17   0  0.0  0.0
18   0  0.0  0.0
19   0  0.0  0.0
       C1   C2   C3
Row1    0  0.0  0.0
Row2    0  0.0  0.0
Row3    0  0.0  0.0
Row4    0  0.0  0.0
Row5    0  0.0  0.0
Row6    0  0.0  0.0
Row7    0  0.0  0.0
Row8    0  0.0  0.0
Row9    0  0.0  0.0
Row10   0  0.0  0.0
Row11   0  0.0  0.0
Row12   0  0.0  0.0
Row13   0  0.0  0.0
Row14   0  0.0  0.0
Row15   0  0.0  0.0
Row16   0  0.0  0.0
Row17   0  0.0  0.0
Row18   0  0.0  0.0
Row19   0  0.0  0.0
Row20   0  0.0  0.0


In [4]:
print(df.iloc[:5])

      C1   C2   C3
Row1   0  0.0  0.0
Row2   0  0.0  0.0
Row3   0  0.0  0.0
Row4   0  0.0  0.0
Row5   0  0.0  0.0


**3.** Create Random integer between 2 to 10 with 4 columns

In [5]:
df = pd.DataFrame(np.random.randint(2,10, size=(2,4)), columns=['A','B','C','D'])
df

A  B  C  D
0  7  9  2  6
1  8  3  8  5

In [6]:
np.random.randint(2,10, size=(2,4))

array([[2, 5, 8, 7],
       [2, 3, 5, 5]])

**4.** Convert a Panda module Series to Python list and it’s type

In [7]:
ds = pd.Series([2,4,6,8,10])
print(ds)
print(type(ds))
print(ds.tolist())
print(type(ds.tolist()))

0     2
1     4
2     6
3     8
4    10
dtype: int64
<class 'pandas.core.series.Series'>
[2, 4, 6, 8, 10]
<class 'list'>


**5.** Convert a dictionary to a Pandas series.

In [8]:
d = {'a': 100, 'b': 200, 'c': 300, 'd': 400, 'e': 800}

pd.Series(d)

a    100
b    200
c    300
d    400
e    800
dtype: int64

**6.** Convert a NumPy array to a Pandas series.

In [9]:
d = np.array([10,20,30,40,50])

pd.Series(d)

0    10
1    20
2    30
3    40
4    50
dtype: int32

**7.** Change the data type of given a column or a Series

In [10]:
s1 = pd.Series(['100', '200', 'python', '300.12', '400'])

# This line applies the pd.to_numeric() function to the Series
# object 's1' with the 'errors' parameter set to 'coerce'. 
# This function attempts to convert each value in the Series 
# object to a numeric type (e.g., integer or float). 
# If a value cannot be converted, 
# it will be replaced with a NaN (not a number) value.
s2 = pd.to_numeric(s1, errors='coerce')
print(s2)

0    100.00
1    200.00
2       NaN
3    300.12
4    400.00
dtype: float64


**8.** Convert the first column of a DataFrame as a Series

In [11]:
d = {'col1': [1, 2, 3, 4, 7, 11], 'col2': [4, 5, 6, 9, 5, 0], 'col3': [7, 5, 8, 12, 1, 11]}
df = pd.DataFrame(data=d)

s1 = df.iloc[:,0]
print(s1)

0     1
1     2
2     3
3     4
4     7
5    11
Name: col1, dtype: int64


**9.** Convert a given Series to an array

In [12]:
s1 = pd.Series(['100', '200', 'python', '300.12', '400'])
type(s1.values)

numpy.ndarray

**10.** Convert Series of lists to one Series

In [13]:
s = pd.Series([
    ['Red', 'Green', 'White'],
    ['Red', 'Black'],
    ['Yellow']])

s = s.apply(pd.Series).stack().reset_index(drop=True)

s

0       Red
1     Green
2     White
3       Red
4     Black
5    Yellow
dtype: object

**11.** Get the items of a given series not present in another given series

In [14]:
sr1 = pd.Series([1, 2, 3, 4, 5])
sr2 = pd.Series([2, 4, 6, 8, 10])

# This line creates a new Pandas Series object 'result' by selecting the elements from the original Series object 'sr1' that are not in the Series object 'sr2'. This is done by applying a boolean mask to the original Series object using the .isin() method to check which elements in 'sr1' are present in 'sr2', and then negating the boolean mask using the tilde (~) operator.
res = sr1[~sr1.isin(sr2)]
res

0    1
2    3
4    5
dtype: int64

**12.** Get the items which are not common of two given series

In [15]:
sr1 = pd.Series([1, 2, 3, 4, 5])
sr2 = pd.Series([2, 4, 6, 8, 10])

sr11 = pd.Series(np.union1d(sr1, sr2))
sr22 = pd.Series(np.intersect1d(sr1, sr2))

res = sr11[~sr11.isin(sr22)]
res

0     1
2     3
4     5
5     6
6     8
7    10
dtype: int64

**13.** Display a summary of the basic information about a specified DataFrame and its data

In [16]:
exam_data  = {'name': ['Anastasia', 'Dima', 'Katherine', 'James', 'Emily', 'Michael', 'Matthew', 'Laura', 'Kevin', 'Jonas'],
        'score': [12.5, 9, 16.5, np.nan, 9, 20, 14.5, np.nan, 8, 19],
        'attempts': [1, 3, 2, 3, 2, 3, 1, 1, 2, 1],
        'qualify': ['yes', 'no', 'yes', 'no', 'no', 'yes', 'yes', 'no', 'no', 'yes']}
labels = ['a', 'b', 'c', 'd', 'e', 'f', 'g', 'h', 'i', 'j']

df = pd.DataFrame(exam_data , index=labels)

df

name  score  attempts qualify
a  Anastasia   12.5         1     yes
b       Dima    9.0         3      no
c  Katherine   16.5         2     yes
d      James    NaN         3      no
e      Emily    9.0         2      no
f    Michael   20.0         3     yes
g    Matthew   14.5         1     yes
h      Laura    NaN         1      no
i      Kevin    8.0         2      no
j      Jonas   19.0         1     yes

**14.** Get the first 3 rows of a given DataFrame.

In [17]:
df.iloc[:3]

name  score  attempts qualify
a  Anastasia   12.5         1     yes
b       Dima    9.0         3      no
c  Katherine   16.5         2     yes

**15.** Select the 'name' and 'score' columns from the following DataFrame.

In [18]:
df[['name','score']]

name  score
a  Anastasia   12.5
b       Dima    9.0
c  Katherine   16.5
d      James    NaN
e      Emily    9.0
f    Michael   20.0
g    Matthew   14.5
h      Laura    NaN
i      Kevin    8.0
j      Jonas   19.0

**16.** Select 'name' and 'score' columns in rows 1, 3, 5, 6 from the following data frame.

In [19]:
df

name  score  attempts qualify
a  Anastasia   12.5         1     yes
b       Dima    9.0         3      no
c  Katherine   16.5         2     yes
d      James    NaN         3      no
e      Emily    9.0         2      no
f    Michael   20.0         3     yes
g    Matthew   14.5         1     yes
h      Laura    NaN         1      no
i      Kevin    8.0         2      no
j      Jonas   19.0         1     yes

In [20]:
df.iloc[[1,3,5,6],[1,3]]

score qualify
b    9.0      no
d    NaN      no
f   20.0     yes
g   14.5     yes

In [21]:
df.loc[['b','d','f','g'],['score','qualify']]

score qualify
b    9.0      no
d    NaN      no
f   20.0     yes
g   14.5     yes

**17.** Select the rows where the number of attempts in the examination is greater than 2.

In [22]:
df[df['attempts']>2]

name  score  attempts qualify
b     Dima    9.0         3      no
d    James    NaN         3      no
f  Michael   20.0         3     yes

**18.** Count the number of rows and columns of a DataFrame.

In [23]:
df.shape

(10, 4)

In [24]:
len(df.axes[0])

10

In [25]:
len(df.axes[1])

4

**19.** Make a Pandas DataFrame with a two-dimensional list

In [26]:
lst=[['food',45],['is',56],['so',1],['delicious',5]]
df = pd.DataFrame(lst, columns=['Name','Number'])
df

Name  Number
0       food      45
1         is      56
2         so       1
3  delicious       5

**20.** Create a DataFrame df from a dictionary data which has the index labels

In [27]:
data = {'animal': ['cat', 'cat', 'snake', 'dog', 'dog', 'cat', 'snake', 'cat', 'dog', 'dog'],
        'age': [2.5, 3, 0.5, np.nan, 5, 2, 4.5, np.nan, 7, 3],
        'visits': [1, 3, 2, 3, 2, 3, 1, 1, 2, 1],
        'priority': ['yes', 'yes', 'no', 'yes', 'no', 'no', 'no', 'yes', 'no', 'no']}

labels = ['a', 'b', 'c', 'd', 'e', 'f', 'g', 'h', 'i', 'j']

df = pd.DataFrame(data, index=labels)

df

animal  age  visits priority
a    cat  2.5       1      yes
b    cat  3.0       3      yes
c  snake  0.5       2       no
d    dog  NaN       3      yes
e    dog  5.0       2       no
f    cat  2.0       3       no
g  snake  4.5       1       no
h    cat  NaN       1      yes
i    dog  7.0       2       no
j    dog  3.0       1       no

**21.** Display a summary of the basic information about this DataFrame and its data

In [28]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 10 entries, a to j
Data columns (total 4 columns):
 #   Column    Non-Null Count  Dtype  
---  ------    --------------  -----  
 0   animal    10 non-null     object 
 1   age       8 non-null      float64
 2   visits    10 non-null     int64  
 3   priority  10 non-null     object 
dtypes: float64(1), int64(1), object(2)
memory usage: 400.0+ bytes


In [29]:
df.describe()

age     visits
count  8.000000  10.000000
mean   3.437500   1.900000
std    2.007797   0.875595
min    0.500000   1.000000
25%    2.375000   1.000000
50%    3.000000   2.000000
75%    4.625000   2.750000
max    7.000000   3.000000

**22.** Return the first 3 rows of the DataFrame df

In [30]:
df.iloc[:3]

animal  age  visits priority
a    cat  2.5       1      yes
b    cat  3.0       3      yes
c  snake  0.5       2       no

**23.** Select just the 'animal' and 'age' columns from the DataFrame df

In [31]:
df.iloc[:3, :2]

animal  age
a    cat  2.5
b    cat  3.0
c  snake  0.5

**24.** Select the data in rows [3, 4, 8] and in columns ['animal', 'age']

In [32]:
df.loc[df.index[[3,4,8]], ['animal', 'age']]

animal  age
d    dog  NaN
e    dog  5.0
i    dog  7.0

**25.** Select only the rows where the number of visits is greater than 3.

In [33]:
df

animal  age  visits priority
a    cat  2.5       1      yes
b    cat  3.0       3      yes
c  snake  0.5       2       no
d    dog  NaN       3      yes
e    dog  5.0       2       no
f    cat  2.0       3       no
g  snake  4.5       1       no
h    cat  NaN       1      yes
i    dog  7.0       2       no
j    dog  3.0       1       no

In [34]:
df[df['visits']>3]

Empty DataFrame
Columns: [animal, age, visits, priority]
Index: []

**26.** Select the rows where the age is missing, i.e. it is NaN.

In [35]:
df[df['age'].isnull()]

animal  age  visits priority
d    dog  NaN       3      yes
h    cat  NaN       1      yes

**27.** Select the rows where the animal is a cat and the age is less than 3.

In [36]:
df[(df['animal']=='cat') & (df['age']<3)]

animal  age  visits priority
a    cat  2.5       1      yes
f    cat  2.0       3       no

**28.** Select the rows the age is between 2 and 4 (inclusive).

In [37]:
df[df['age'].between(2,4)]

animal  age  visits priority
a    cat  2.5       1      yes
b    cat  3.0       3      yes
f    cat  2.0       3       no
j    dog  3.0       1       no

**29.** Change the age in row 'f' to 1.5.

In [38]:
df.loc['f','age'] = 1.5
df

animal  age  visits priority
a    cat  2.5       1      yes
b    cat  3.0       3      yes
c  snake  0.5       2       no
d    dog  NaN       3      yes
e    dog  5.0       2       no
f    cat  1.5       3       no
g  snake  4.5       1       no
h    cat  NaN       1      yes
i    dog  7.0       2       no
j    dog  3.0       1       no

**30.** Calculate the sum of all visits in df (i.e. find the total number of visits).

In [39]:
df['visits'].sum()

19

**31.** Calculate the mean age for each different animal in df.

In [40]:
df.groupby('animal')['age'].sum()

animal
cat       7.0
dog      15.0
snake     5.0
Name: age, dtype: float64

**32.** Append a new row 'k' to df with your choice of values for each column. Then delete that row to return the original DataFrame.

In [41]:
# adding thr row

df.loc['k']=[5.5,'dog','no',2]
df

animal  age visits priority
a    cat  2.5      1      yes
b    cat  3.0      3      yes
c  snake  0.5      2       no
d    dog  NaN      3      yes
e    dog  5.0      2       no
f    cat  1.5      3       no
g  snake  4.5      1       no
h    cat  NaN      1      yes
i    dog  7.0      2       no
j    dog  3.0      1       no
k    5.5  dog     no        2

In [42]:
df.drop('k', inplace=True)
df

animal  age visits priority
a    cat  2.5      1      yes
b    cat  3.0      3      yes
c  snake  0.5      2       no
d    dog  NaN      3      yes
e    dog  5.0      2       no
f    cat  1.5      3       no
g  snake  4.5      1       no
h    cat  NaN      1      yes
i    dog  7.0      2       no
j    dog  3.0      1       no

**33.** Count the number of each type of animal in df.

In [43]:
df['animal'].value_counts()

animal
cat      4
dog      4
snake    2
Name: count, dtype: int64

**34.** Sort df first by the values in the 'age' in descending order, then by the value in the 'visits' column in ascending order (so row i should be first, and row d should be last).

In [44]:
df.sort_values(by=['age','visits'], ascending=[False,True])

animal  age visits priority
i    dog  7.0      2       no
e    dog  5.0      2       no
g  snake  4.5      1       no
j    dog  3.0      1       no
b    cat  3.0      3      yes
a    cat  2.5      1      yes
f    cat  1.5      3       no
c  snake  0.5      2       no
h    cat  NaN      1      yes
d    dog  NaN      3      yes

**35.** The 'priority' column contains the values 'yes' and 'no'. Replace this column with a column of boolean values: 'yes' should be True and 'no' should be False.

In [45]:
df['priority'] = df['priority'].map({'yes':True, 'no':False})
df

animal  age visits  priority
a    cat  2.5      1      True
b    cat  3.0      3      True
c  snake  0.5      2     False
d    dog  NaN      3      True
e    dog  5.0      2     False
f    cat  1.5      3     False
g  snake  4.5      1     False
h    cat  NaN      1      True
i    dog  7.0      2     False
j    dog  3.0      1     False

**36.** In the 'animal' column, change the 'snake' entries to 'python'.

In [46]:
df['animal'] = df['animal'].replace('snake', 'python')
df

animal  age visits  priority
a     cat  2.5      1      True
b     cat  3.0      3      True
c  python  0.5      2     False
d     dog  NaN      3      True
e     dog  5.0      2     False
f     cat  1.5      3     False
g  python  4.5      1     False
h     cat  NaN      1      True
i     dog  7.0      2     False
j     dog  3.0      1     False

**37.** For each animal type and each number of visits, find the mean age. In other words, each row is an animal, each column is a number of visits and the values are the mean ages (hint: use a pivot table)

In [47]:
pd.pivot_table(df, values="age", index='animal', 
               columns=['visits'], aggfunc="mean")

visits    1    2     3
animal                
cat     2.5  NaN  2.25
dog     3.0  6.0   NaN
python  4.5  0.5   NaN

**38.** You have a DataFrame df with a column 'A' of integers. For example:

df = pd.DataFrame({'A': [1, 2, 2, 3, 4, 5, 5, 5, 6, 7, 7]})
How do you filter out rows which contain the same integer as the row immediately above?

You should be left with a column containing the following values:

1, 2, 3, 4, 5, 6, 7

In [48]:
df = pd.DataFrame({'A': [1, 2, 2, 3, 4, 5, 5, 5, 6, 7, 7]})
df

A
0   1
1   2
2   2
3   3
4   4
5   5
6   5
7   5
8   6
9   7
10  7

In [49]:
df.loc[df['A'].shift() != df['A']]

A
0  1
1  2
3  3
4  4
5  5
8  6
9  7

**39.** Given a DataFrame of numeric values, say

df = pd.DataFrame(np.random.random(size=(5, 3))) # a 5x3 frame of float values
how do you subtract the row mean from each element in the row?

In [50]:
df = pd.DataFrame(np.random.random(size=(5, 3)))
df

0         1         2
0  0.048295  0.215328  0.689766
1  0.496419  0.013474  0.030404
2  0.875194  0.991913  0.912881
3  0.807116  0.553109  0.163045
4  0.139133  0.329349  0.004447

In [51]:
df.sub(df.mean(axis=1), axis=0)

0         1         2
0 -0.269502 -0.102468  0.371970
1  0.316320 -0.166625 -0.149695
2 -0.051469  0.065250 -0.013782
3  0.299359  0.045352 -0.344711
4 -0.018510  0.171706 -0.153196

In [52]:
df.mean(axis=1)

0    0.317796
1    0.180099
2    0.926663
3    0.507757
4    0.157643
dtype: float64

**40.** Suppose you have DataFrame with 10 columns of real numbers, for example:

df = pd.DataFrame(np.random.random(size=(5, 10)), columns=list('abcdefghij'))
Which column of numbers has the smallest sum? Return that column's label.

In [53]:
df = pd.DataFrame(np.random.random(size=(5, 10)), columns=list('abcdefghij'))
df

a         b         c         d         e         f         g  \
0  0.834297  0.260767  0.288602  0.895100  0.138922  0.929253  0.500533   
1  0.804782  0.697778  0.808906  0.815876  0.559086  0.765349  0.210467   
2  0.637117  0.754570  0.197969  0.384406  0.760020  0.246587  0.995990   
3  0.476717  0.397538  0.198137  0.891440  0.315529  0.645292  0.503409   
4  0.741975  0.208436  0.460811  0.198786  0.494356  0.031725  0.715665   

          h         i         j  
0  0.580848  0.125232  0.088049  
1  0.492318  0.547801  0.886019  
2  0.146197  0.293923  0.199575  
3  0.033890  0.240092  0.649588  
4  0.935084  0.276961  0.225587

In [54]:
df.sum().idxmin()

'i'

**41.** How do you count how many unique rows a DataFrame has (i.e. ignore all rows that are duplicates)? As input, use a DataFrame of zeros and ones with 10 rows and 3 columns.

df = pd.DataFrame(np.random.randint(0, 2, size=(10, 3)))

In [55]:
df = pd.DataFrame(np.random.randint(0, 2, size=(10, 3)))
df

0  1  2
0  1  0  1
1  0  1  1
2  0  1  0
3  1  1  0
4  1  1  1
5  0  1  0
6  1  1  1
7  0  1  1
8  0  0  0
9  1  0  1

In [56]:
len(df.drop_duplicates(keep=False))

2

**42.** In the cell below, you have a DataFrame df that consists of 10 columns of floating-point numbers. Exactly 5 entries in each row are NaN values.

For each row of the DataFrame, find the column which contains the third NaN value.

You should return a Series of column labels: e, c, d, h, d

In [57]:
nan = np.nan

data = [[0.04,  nan,  nan, 0.25,  nan, 0.43, 0.71, 0.51,  nan,  nan],
        [ nan,  nan,  nan, 0.04, 0.76,  nan,  nan, 0.67, 0.76, 0.16],
        [ nan,  nan, 0.5 ,  nan, 0.31, 0.4 ,  nan,  nan, 0.24, 0.01],
        [0.49,  nan,  nan, 0.62, 0.73, 0.26, 0.85,  nan,  nan,  nan],
        [ nan,  nan, 0.41,  nan, 0.05,  nan, 0.61,  nan, 0.48, 0.68]]

columns = list('abcdefghij')

df = pd.DataFrame(data, columns=columns)
# write a solution to the question here

In [58]:
df

a   b     c     d     e     f     g     h     i     j
0  0.04 NaN   NaN  0.25   NaN  0.43  0.71  0.51   NaN   NaN
1   NaN NaN   NaN  0.04  0.76   NaN   NaN  0.67  0.76  0.16
2   NaN NaN  0.50   NaN  0.31  0.40   NaN   NaN  0.24  0.01
3  0.49 NaN   NaN  0.62  0.73  0.26  0.85   NaN   NaN   NaN
4   NaN NaN  0.41   NaN  0.05   NaN  0.61   NaN  0.48  0.68

In [59]:
(df.isnull().cumsum(axis=1)==3).idxmax(axis=1)

# explanation
# 1. df.isnull()
# Converts the DataFrame into Boolean values (True if the element is NaN, otherwise False).

# 2. cumsum(axis=1)
# Computes cumulative sum horizontally (across columns).
# Each row now tracks the running total of NaNs from left to right.

# 3. == 3
# Creates a Boolean DataFrame marking positions where the cumulative sum reaches exactly 3 (third NaN occurrence).

# 4.idxmax(axis=1)
# Returns the column label of the first occurrence of the True value for each row.

0    e
1    c
2    d
3    h
4    d
dtype: object

**43.** A DataFrame has a column of groups 'grps' and and column of integer values 'vals': 

```python
df = pd.DataFrame({'grps': list('aaabbcaabcccbbc'), 
                   'vals': [12,345,3,1,45,14,4,52,54,23,235,21,57,3,87]})
```
For each *group*, find the sum of the three greatest values. You should end up with the answer as follows:
```
grps
a    409
b    156
c    345
```

In [60]:
df = pd.DataFrame({'grps': list('aaabbcaabcccbbc'), 
                   'vals': [12,345,3,1,45,14,4,52,54,23,235,21,57,3,87]})
df
# write a solution to the question here

grps  vals
0     a    12
1     a   345
2     a     3
3     b     1
4     b    45
5     c    14
6     a     4
7     a    52
8     b    54
9     c    23
10    c   235
11    c    21
12    b    57
13    b     3
14    c    87

In [61]:
df.groupby('grps')['vals'].apply(lambda x: x.nlargest(3).sum())

# explanation
# - groupby('grps'): Group rows by the column 'grps'.

# - x.nlargest(3): For each group, get the three largest values.

# - sum(): Sum these top three values for each group.

grps
a    409
b    156
c    345
Name: vals, dtype: int64

**44.** The DataFrame `df` constructed below has two integer columns 'A' and 'B'. The values in 'A' are between 1 and 100 (inclusive). 

For each group of 10 consecutive integers in 'A' (i.e. `(0, 10]`, `(10, 20]`, ...), calculate the sum of the corresponding values in column 'B'.

The answer should be a Series as follows:

```
A
(0, 10]      635
(10, 20]     360
(20, 30]     315
(30, 40]     306
(40, 50]     750
(50, 60]     284
(60, 70]     424
(70, 80]     526
(80, 90]     835
(90, 100]    852
```

In [62]:
df = pd.DataFrame(np.random.RandomState(8765).randint(1, 101, size=(100, 2)), columns = ["A", "B"])

df.groupby(pd.cut(df['A'], np.arange(0, 101, 10)))['B'].sum()
# Explanation:
# df.groupby(pd.cut(df['A'], np.arange(0, 101, 10)))
# Takes column 'A' and assigns each value to intervals such as (0,10], (10,20], (20,30], etc.

# ['B'].sum()
# Once grouped, this calculates the sum of column 'B' for each interval group identified by column 'A'.

C:\Users\ASUS\AppData\Local\Temp\ipykernel_8812\1704753638.py:3: FutureWarning: The default of observed=False is deprecated and will be changed to True in a future version of pandas. Pass observed=False to retain current behavior or observed=True to adopt the future default and silence this warning.
  df.groupby(pd.cut(df['A'], np.arange(0, 101, 10)))['B'].sum()


A
(0, 10]      635
(10, 20]     360
(20, 30]     315
(30, 40]     306
(40, 50]     750
(50, 60]     284
(60, 70]     424
(70, 80]     526
(80, 90]     835
(90, 100]    852
Name: B, dtype: int32

**45.** Consider a DataFrame `df` where there is an integer column 'X':
```python
df = pd.DataFrame({'X': [7, 2, 0, 3, 4, 2, 5, 0, 3, 4]})
```
For each value, count the difference back to the previous zero (or the start of the Series, whichever is closer). These values should therefore be 

```
[1, 2, 0, 1, 2, 3, 4, 0, 1, 2]
```

Make this a new column 'Y'.

In [63]:
df = pd.DataFrame({'X': [7, 2, 0, 3, 4, 2, 5, 0, 3, 4]})
zero_pos = (df['X'] == 0)
print(zero_pos)
print("**********************")

# creates groups between zeros.
groups = zero_pos.cumsum()
print(groups)
print("**********************")

# .cumcount() counts occurrences within each group.
cum_sum = df.groupby(groups).cumcount()
print(cum_sum)
print("**********************")

# (~zero_positions) adds 1 to all positions except zeros to match the required output exactly.
df['Y'] = cum_sum + (~zero_pos)
df

0    False
1    False
2     True
3    False
4    False
5    False
6    False
7     True
8    False
9    False
Name: X, dtype: bool
**********************
0    0
1    0
2    1
3    1
4    1
5    1
6    1
7    2
8    2
9    2
Name: X, dtype: int32
**********************
0    0
1    1
2    0
3    1
4    2
5    3
6    4
7    0
8    1
9    2
dtype: int64
**********************


X  Y
0  7  1
1  2  2
2  0  0
3  3  2
4  4  3
5  2  4
6  5  5
7  0  0
8  3  2
9  4  3

**46.** You are given the DataFrame below with a column of group IDs, 'grps', and a column of corresponding integer values, 'vals'.

```python
df = pd.DataFrame({"vals": np.random.RandomState(31).randint(-30, 30, size=15), 
                   "grps": np.random.RandomState(31).choice(["A", "B"], 15)})
```

Create a new column 'patched_values' which contains the same values as the 'vals' any negative values in 'vals' with the group mean:

```
    vals grps  patched_vals
0    -12    A          13.6
1     -7    B          28.0
2    -14    A          13.6
3      4    A           4.0
4     -7    A          13.6
5     28    B          28.0
6     -2    A          13.6
7     -1    A          13.6
8      8    A           8.0
9     -2    B          28.0
10    28    A          28.0
11    12    A          12.0
12    16    A          16.0
13   -24    A          13.6
14   -12    A          13.6
```

In [64]:
df = pd.DataFrame({"vals": np.random.RandomState(31).randint(-30, 30, size=15), 
                   "grps": np.random.RandomState(31).choice(["A", "B"], 15)})

# Calculate mean of non-negative values per group
# .transform() ensures alignment with original DataFrame.
# Compute the mean of non-negative values for each group ('A', 'B') using:
group_means = df.groupby('grps')['vals'].transform(lambda x: x[x >= 0].mean())

# Replace negative values with the corresponding group's mean
df['patched_values'] = np.where(df['vals'] < 0, group_means, df['vals'])

print(df)

    vals grps  patched_values
0    -12    A            13.6
1     -7    B            28.0
2    -14    A            13.6
3      4    A             4.0
4     -7    A            13.6
5     28    B            28.0
6     -2    A            13.6
7     -1    A            13.6
8      8    A             8.0
9     -2    B            28.0
10    28    A            28.0
11    12    A            12.0
12    16    A            16.0
13   -24    A            13.6
14   -12    A            13.6


**47.** Implement a rolling mean over groups with window size 3, which ignores NaN value. For example consider the following DataFrame:

```python
>>> df = pd.DataFrame({'group': list('aabbabbbabab'),
                       'value': [1, 2, 3, np.nan, 2, 3, np.nan, 1, 7, 3, np.nan, 8]})
>>> df
   group  value
0      a    1.0
1      a    2.0
2      b    3.0
3      b    NaN
4      a    2.0
5      b    3.0
6      b    NaN
7      b    1.0
8      a    7.0
9      b    3.0
10     a    NaN
11     b    8.0
```
The goal is to compute the Series:

```
0     1.000000
1     1.500000
2     3.000000
3     3.000000
4     1.666667
5     3.000000
6     3.000000
7     2.000000
8     3.666667
9     2.000000
10    4.500000
11    4.000000
```
E.g. the first window of size three for group 'b' has values 3.0, NaN and 3.0 and occurs at row index 5. Instead of being NaN the value in the new column at this row index should be 3.0 (just the two non-NaN values are used to compute the mean (3+3)/2)

In [65]:
df = pd.DataFrame({'group': list('aabbabbbabab'),
                       'value': [1, 2, 3, np.nan, 2, 3, np.nan, 1, 7, 3, np.nan, 8]})


# Compute rolling mean by group with window=3, ignoring NaNs
df['rolling_mean'] = df.groupby('group')['value'].transform(
    lambda x: x.rolling(window=3, min_periods=1).mean()
)
df

group  value  rolling_mean
0      a    1.0      1.000000
1      a    2.0      1.500000
2      b    3.0      3.000000
3      b    NaN      3.000000
4      a    2.0      1.666667
5      b    3.0      3.000000
6      b    NaN      3.000000
7      b    1.0      2.000000
8      a    7.0      3.666667
9      b    3.0      2.000000
10     a    NaN      4.500000
11     b    8.0      4.000000

**Explanation:**
<br>
🔹 df.groupby('group')['value']
Groups the data by the values in the 'group' column ('a' and 'b'), and selects the 'value' column for each group.

So you get two groups:

Group 'a': values = [1.0, 2.0, 2.0, 7.0, NaN]

Group 'b': values = [3.0, NaN, 3.0, NaN, 1.0, 3.0, 8.0]

🔹 .transform(lambda x: ...)
Applies the following function to each group, and aligns the result back to the original DataFrame (maintains row order).

🔹 .rolling(window=3, min_periods=1).mean()
- window=3: Computes the average over the current and last 2 values (i.e. rolling over 3 rows).

- min_periods=1: Ensures the mean is computed as long as there is at least 1 valid (non-NaN) value in the window.

- It ignores NaN values by default in .mean().

So for each group:

At every row, it looks back at the last 3 values (including current row), and computes the mean while skipping NaNs.

🅰️ **Group 'a' — Row indexes: 0, 1, 4, 8, 10**
**Row 0 (index 0)**
Last 1 value: [1.0] → mean = 1.0

**Row 1 (index 1)**
Last 2 values: [1.0, 2.0] → mean = 1.5

**Row 4 (index 4)**
Last 3 values: [1.0, 2.0, 2.0] → mean = 1.6667

**Row 8 (index 8)**
Last 3 values: [2.0, 2.0, 7.0] → mean = 3.6667

**Row 10 (index 10)**
Last 3 values: [7.0, NaN, NaN] → mean of non-NaNs = 7.0 → rolling window: [2.0, 7.0, NaN] → mean = 4.5

🅱️ **Group 'b' — Row indexes: 2, 3, 5, 6, 7, 9, 11**
**Row 2 (index 2)**
[3.0] → mean = 3.0

**Row 3 (index 3)**
[3.0, NaN] → mean = 3.0 → window = [3.0, NaN] → 3.0

**Row 5 (index 5)**
[3.0, NaN, 3.0] → mean = (3+3)/2 = 3.0

**Row 6 (index 6)**
[NaN, 3.0, NaN] → only one valid: [3.0] → mean = 3.0

**Row 7 (index 7)**
[3.0, NaN, 1.0] → mean = (3+1)/2 = 2.0

**Row 9 (index 9)**
[NaN, 1.0, 3.0] → mean = (1+3)/2 = 2.0

**Row 11 (index 11)**
[1.0, 3.0, 8.0] → mean = (1+3+8)/3 = 4.0

**END OF EXPLANATION**

## Cleaning Data

### Making a DataFrame easier to work with

Difficulty: *easy/medium*

It happens all the time: someone gives you data containing malformed strings, Python, lists and missing data. How do you tidy it up so you can get on with the analysis?

Take this monstrosity as the DataFrame to use in the following puzzles:

```python
df = pd.DataFrame({'From_To': ['LoNDon_paris', 'MAdrid_miLAN', 'londON_StockhOlm', 
                               'Budapest_PaRis', 'Brussels_londOn'],
              'FlightNumber': [10045, np.nan, 10065, np.nan, 10085],
              'RecentDelays': [[23, 47], [], [24, 43, 87], [13], [67, 32]],
                   'Airline': ['KLM(!)', '<Air France> (12)', '(British Airways. )', 
                               '12. Air France', '"Swiss Air"']})
```
Formatted, it looks like this:

```
            From_To  FlightNumber  RecentDelays              Airline
0      LoNDon_paris       10045.0      [23, 47]               KLM(!)
1      MAdrid_miLAN           NaN            []    <Air France> (12)
2  londON_StockhOlm       10065.0  [24, 43, 87]  (British Airways. )
3    Budapest_PaRis           NaN          [13]       12. Air France
4   Brussels_londOn       10085.0      [67, 32]          "Swiss Air"
```


(It's some flight data I made up; it's not meant to be accurate in any way.)

**48.** Some values in the the FlightNumber column are missing (they are NaN). These numbers are meant to increase by 10 with each row so 10055 and 10075 need to be put in place. Modify df to fill in these missing numbers and make the column an integer column (instead of a float column).

In [66]:
df = pd.DataFrame({'From_To': ['LoNDon_paris', 'MAdrid_miLAN', 'londON_StockhOlm', 
                               'Budapest_PaRis', 'Brussels_londOn'],
              'FlightNumber': [10045, np.nan, 10065, np.nan, 10085],
              'RecentDelays': [[23, 47], [], [24, 43, 87], [13], [67, 32]],
                   'Airline': ['KLM(!)', '<Air France> (12)', '(British Airways. )', 
                               '12. Air France', '"Swiss Air"']})
df

From_To  FlightNumber  RecentDelays              Airline
0      LoNDon_paris       10045.0      [23, 47]               KLM(!)
1      MAdrid_miLAN           NaN            []    <Air France> (12)
2  londON_StockhOlm       10065.0  [24, 43, 87]  (British Airways. )
3    Budapest_PaRis           NaN          [13]       12. Air France
4   Brussels_londOn       10085.0      [67, 32]          "Swiss Air"

In [67]:
df['FlightNumber'] = df['FlightNumber'].interpolate().astype(int)
df

From_To  FlightNumber  RecentDelays              Airline
0      LoNDon_paris         10045      [23, 47]               KLM(!)
1      MAdrid_miLAN         10055            []    <Air France> (12)
2  londON_StockhOlm         10065  [24, 43, 87]  (British Airways. )
3    Budapest_PaRis         10075          [13]       12. Air France
4   Brussels_londOn         10085      [67, 32]          "Swiss Air"

In [68]:
df

From_To  FlightNumber  RecentDelays              Airline
0      LoNDon_paris         10045      [23, 47]               KLM(!)
1      MAdrid_miLAN         10055            []    <Air France> (12)
2  londON_StockhOlm         10065  [24, 43, 87]  (British Airways. )
3    Budapest_PaRis         10075          [13]       12. Air France
4   Brussels_londOn         10085      [67, 32]          "Swiss Air"

**49.** The From_To column would be better as two separate columns! Split each string on the underscore delimiter _ to give a new temporary DataFrame called 'temp' with the correct values. Assign the correct column names 'From' and 'To' to this temporary DataFrame. Also Notice how the capitalisation of the city names is all mixed up in this temporary DataFrame 'temp'. Standardise the strings so that only the first letter is uppercase (e.g. "londON" should become "London".)

In [69]:
df[['From', 'To']] = df['From_To'].str.split('_',expand=True)
df['From'] = df['From'].str.title()
df['To'] = df['To'].str.title()
df

From_To  FlightNumber  RecentDelays              Airline  \
0      LoNDon_paris         10045      [23, 47]               KLM(!)   
1      MAdrid_miLAN         10055            []    <Air France> (12)   
2  londON_StockhOlm         10065  [24, 43, 87]  (British Airways. )   
3    Budapest_PaRis         10075          [13]       12. Air France   
4   Brussels_londOn         10085      [67, 32]          "Swiss Air"   

       From         To  
0    London      Paris  
1    Madrid      Milan  
2    London  Stockholm  
3  Budapest      Paris  
4  Brussels     London

In [70]:
df.drop('From_To', axis=1, inplace=True)
df

FlightNumber  RecentDelays              Airline      From         To
0         10045      [23, 47]               KLM(!)    London      Paris
1         10055            []    <Air France> (12)    Madrid      Milan
2         10065  [24, 43, 87]  (British Airways. )    London  Stockholm
3         10075          [13]       12. Air France  Budapest      Paris
4         10085      [67, 32]          "Swiss Air"  Brussels     London

**50.** In the Airline column, you can see some extra puctuation and symbols have appeared around the airline names. Pull out just the airline name. E.g. '(British Airways. )' should become 'British Airways'.

In [71]:
df['Airline'] = df['Airline'].str.replace(r'[^a-zA-Z\s]', '', regex=True).str.strip()
df

FlightNumber  RecentDelays          Airline      From         To
0         10045      [23, 47]              KLM    London      Paris
1         10055            []       Air France    Madrid      Milan
2         10065  [24, 43, 87]  British Airways    London  Stockholm
3         10075          [13]       Air France  Budapest      Paris
4         10085      [67, 32]        Swiss Air  Brussels     London

**51.** In the RecentDelays column, the values have been entered into the DataFrame as a list. We would like each first value in its own column, each second value in its own column, and so on. If there isn't an Nth value, the value should be NaN.

Expand the Series of lists into a DataFrame named delays, rename the columns delay_1, delay_2, etc. and replace the unwanted RecentDelays column in df with delays.

In [72]:
df_new = df.copy(deep=True)
delays = df_new['RecentDelays'].apply(pd.Series)
delays.columns = [f'delay_{i+1}' for i in delays.columns]
df_new = pd.concat([df_new, delays], axis=1)
df_new

FlightNumber  RecentDelays          Airline      From         To  delay_1  \
0         10045      [23, 47]              KLM    London      Paris     23.0   
1         10055            []       Air France    Madrid      Milan      NaN   
2         10065  [24, 43, 87]  British Airways    London  Stockholm     24.0   
3         10075          [13]       Air France  Budapest      Paris     13.0   
4         10085      [67, 32]        Swiss Air  Brussels     London     67.0   

   delay_2  delay_3  
0     47.0      NaN  
1      NaN      NaN  
2     43.0     87.0  
3      NaN      NaN  
4     32.0      NaN

In [73]:
df_new.drop('RecentDelays', axis=1, inplace=True)
df_new

FlightNumber          Airline      From         To  delay_1  delay_2  \
0         10045              KLM    London      Paris     23.0     47.0   
1         10055       Air France    Madrid      Milan      NaN      NaN   
2         10065  British Airways    London  Stockholm     24.0     43.0   
3         10075       Air France  Budapest      Paris     13.0      NaN   
4         10085        Swiss Air  Brussels     London     67.0     32.0   

   delay_3  
0      NaN  
1      NaN  
2     87.0  
3      NaN  
4      NaN

## Using MultiIndexes

**52.** Given the lists letters = ['A', 'B', 'C'] and numbers = list(range(10)), construct a MultiIndex object from the product of the two lists. Use it to index a Series of random numbers. Call this Series s.

In [74]:
letters = ['A', 'B', 'C']
numbers = list(range(10))

# Step 1: Create a MultiIndex from the product of the two lists
multi_index = pd.MultiIndex.from_product([letters, numbers], names=['Letter', 'Number'])
multi_index

MultiIndex([('A', 0),
            ('A', 1),
            ('A', 2),
            ('A', 3),
            ('A', 4),
            ('A', 5),
            ('A', 6),
            ('A', 7),
            ('A', 8),
            ('A', 9),
            ('B', 0),
            ('B', 1),
            ('B', 2),
            ('B', 3),
            ('B', 4),
            ('B', 5),
            ('B', 6),
            ('B', 7),
            ('B', 8),
            ('B', 9),
            ('C', 0),
            ('C', 1),
            ('C', 2),
            ('C', 3),
            ('C', 4),
            ('C', 5),
            ('C', 6),
            ('C', 7),
            ('C', 8),
            ('C', 9)],
           names=['Letter', 'Number'])

In [77]:
s = pd.Series(np.random.randn(len(multi_index)), index=multi_index)
s
# Explanation:
# np.random.randn(n) generates n random numbers from a standard normal distribution (mean=0, std=1).
# len(multi_index) tells you how many total combinations there are (in this case, 3 * 10 = 30).

Letter  Number
A       0         1.195414
        1        -0.346384
        2         0.710760
        3        -0.810707
        4         0.283131
        5        -0.893565
        6         0.880194
        7         0.066425
        8         0.148851
        9        -0.324581
B       0        -2.055239
        1         1.362094
        2        -0.496943
        3        -1.047739
        4         1.550024
        5         0.948727
        6        -0.278573
        7         0.430243
        8        -0.724240
        9        -1.477983
C       0        -0.962775
        1         0.881106
        2         0.213212
        3        -1.271780
        4         0.537178
        5        -0.124388
        6         1.953340
        7        -1.488654
        8        -0.494559
        9         0.754861
dtype: float64

**53.** Check the index of s is lexicographically sorted (this is a necessary proprty for indexing to work correctly with a MultiIndex).

In [79]:
# ✅ Check if the index is lexicographically sorted
is_sorted = s.index.is_monotonic_increasing

print("Is the MultiIndex lexicographically sorted?", is_sorted)
is_sorted

Is the MultiIndex lexicographically sorted? True


True

**54.** Select the labels 1, 3 and 6 from the second level of the MultiIndexed Series.

In [84]:
s.loc[:,[1,3,6]]

Letter  Number
A       1        -0.346384
        3        -0.810707
        6         0.880194
B       1         1.362094
        3        -1.047739
        6        -0.278573
C       1         0.881106
        3        -1.271780
        6         1.953340
dtype: float64

**55.** Slice the Series s; slice up to label 'B' for the first level and from label 5 onwards for the second level.

In [88]:
s.loc[:'B',5:]

Letter  Number
A       5        -0.893565
        6         0.880194
        7         0.066425
        8         0.148851
        9        -0.324581
B       5         0.948727
        6        -0.278573
        7         0.430243
        8        -0.724240
        9        -1.477983
dtype: float64

**56.** Sum the values in s for each label in the first level (you should have Series giving you a total for labels A, B and C).

In [93]:
s.groupby('Letter').sum()

Letter
A    0.909538
B   -1.789629
C   -0.002459
dtype: float64